tests/python/unittest/test_tir_structural_equal.py

```diff
@@ -55,8 +55,8 @@ def test_exprs():
     assert tvm.ir.structural_equal(zx * zx, zx * zx)
     assert tvm.ir.structural_equal(zx * zx, zy * zy, map_free_vars=True)
     assert not tvm.ir.structural_equal(zx * zx, zy * zy, map_free_vars=False)
-    assert not tvm.ir.structural_equal(zx * zx, (vx + vx) * (vx + vx),
-                                       map_free_vars=False)
+    assert tvm.ir.structural_equal(zx * zx, (vx + vx) * (vx + vx),
```

Comment:
This doesnt seems very right to me. What if vx has effect?

Reply:
if vx has effect, then the effect itself should be a graph node(in the case of zx) and the equality will no longer holds.

Reply:
Note that the main point of difference is whether we treat a node as graph node or normal one. the current semantics on the TIR is to treat expr nodes as normal nodes. We could change the definition by marking all TIR expr as graph nodes as well, which will give the other behavior.

Reply:
BTW, in this case, vx is a variable, so this line alone won't have any effect(other statements that generates bindings to vx might have effects)

Reply:
I see, make sense to me.

Reply:
I'm also confused by this one, I don't see the argument for why these are ever equal.

Reply:
note that zx are short hand for `(vx+vx)` here, they are not separate variable. So the comparison  here is a comparing a diamond dag of (x+x) * (x+x) vs a tree. In the case of TIR both are simply treated as a tree atm